File: lib/screens/daily_stats_screen.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:intl/intl.dart';
import '../providers/stats_provider.dart';
import '../providers/user_provider.dart';
import '../data/focus_session.dart';
import '../data/daily_event.dart';

class DailyStatsScreen extends StatefulWidget {
  final DateTime initialDate;
  const DailyStatsScreen({super.key, required this.initialDate});

  @override
  State<DailyStatsScreen> createState() => _DailyStatsScreenState();
}

class _DailyStatsScreenState extends State<DailyStatsScreen> {
  late DateTime _currentDate;

  @override
  void initState() {
    super.initState();
    _currentDate = widget.initialDate;
  }

  @override
  Widget build(BuildContext context) {
    final statsProvider = context.watch<StatsProvider>();
    final userProvider = context.watch<UserProvider>();
    final user = userProvider.user;

    final dailySessions = statsProvider.getSessionsForDay(_currentDate);
    final totalFocussedMillis = dailySessions.fold(0, (sum, s) => sum + s.duration);
    final dailyTargetMillis = user?.dailyFocusTarget ?? 7200000;
    final progress = totalFocussedMillis / dailyTargetMillis;

    return Scaffold(
      appBar: AppBar(
        title: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            IconButton(
              icon: const Icon(Icons.arrow_back_ios, size: 16),
              onPressed: () => setState(() => _currentDate = _currentDate.subtract(const Duration(days: 1))),
            ),
            Text(DateFormat('EEEE, MMM d').format(_currentDate), style: const TextStyle(fontSize: 16)),
            IconButton(
              icon: const Icon(Icons.arrow_forward_ios, size: 16),
              onPressed: () => setState(() => _currentDate = _currentDate.add(const Duration(days: 1))),
            ),
          ],
        ),
      ),
      body: FutureBuilder<List<DailyEvent>>(
        future: statsProvider.getEventsForDay(_currentDate),
        builder: (context, snapshot) {
          final events = snapshot.data ?? [];
          final timelineItems = [...dailySessions, ...events]..sort((a, b) {
            final timeA = a is FocusSession ? a.startTime : (a as DailyEvent).timestamp;
            final timeB = b is FocusSession ? b.startTime : (b as DailyEvent).timestamp;
            return timeA.compareTo(timeB);
          });

          return ListView(
            padding: const EdgeInsets.all(16),
            children: [
              _Circular24hGraph(sessions: dailySessions),
              const SizedBox(height: 32),
              Column(
                children: [
                  Text(
                    '${totalFocussedMillis ~/ 60000} / ${dailyTargetMillis ~/ 60000} mins',
                    style: Theme.of(context).textTheme.titleLarge?.copyWith(fontWeight: FontWeight.bold),
                  ),
                  const SizedBox(height: 8),
                  LinearProgressIndicator(
                    value: progress.clamp(0.0, 1.0),
                    minHeight: 8,
                    borderRadius: BorderRadius.circular(4),
                  ),
                ],
              ),
              const Divider(height: 48),
              Text('Timeline', style: Theme.of(context).textTheme.titleMedium),
              const SizedBox(height: 16),
              ...timelineItems.map((item) {
                if (item is FocusSession) {
                  return _TimelineSessionItem(session: item);
                } else {
                  return _TimelineEventItem(event: item as DailyEvent);
                }
              }),
            ],
          );
        },
      ),
    );
  }
}

class _Circular24hGraph extends StatelessWidget {
  final List<FocusSession> sessions;
  const _Circular24hGraph({required this.sessions});

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Stack(
        alignment: Alignment.center,
        children: [
          SizedBox(
            width: 250,
            height: 250,
            child: CustomPaint(
              painter: _Circular24hPainter(
                sessions: sessions,
                primaryColor: Theme.of(context).colorScheme.primary,
                secondaryColor: Theme.of(context).colorScheme.surfaceContainerHighest,
              ),
            ),
          ),
          Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              Text('24h', style: Theme.of(context).textTheme.headlineMedium?.copyWith(fontWeight: FontWeight.bold)),
              Text('Focus Distribution', style: Theme.of(context).textTheme.bodySmall),
            ],
          ),
        ],
      ),
    );
  }
}

class _Circular24hPainter extends CustomPainter {
  final List<FocusSession> sessions;
  final Color primaryColor;
  final Color secondaryColor;

  _Circular24hPainter({
    required this.sessions,
    required this.primaryColor,
    required this.secondaryColor,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final center = Offset(size.width / 2, size.height / 2);
    final radius = size.shortestSide / 2;
    const strokeWidth = 20.0;

    final basePaint = Paint()
      ..color = secondaryColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth;

    canvas.drawCircle(center, radius - strokeWidth / 2, basePaint);

    // Draw hour marks
    final markPaint = Paint()
      ..color = Colors.white.withValues(alpha: 0.5)
      ..strokeWidth = 2;

    for (int i = 0; i < 24; i++) {
      final angle = (i * 15 - 90) * pi / 180;
      final inner = Offset(
        center.dx + (radius - strokeWidth) * cos(angle),
        center.dy + (radius - strokeWidth) * sin(angle),
      );
      final outer = Offset(
        center.dx + radius * cos(angle),
        center.dy + radius * sin(angle),
      );
      canvas.drawLine(inner, outer, markPaint);
    }

    // Draw sessions
    final sessionPaint = Paint()
      ..color = primaryColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth
      ..strokeCap = StrokeCap.round;

    for (var session in sessions) {
      final startDate = DateTime.fromMillisecondsSinceEpoch(session.startTime);
      final endDate = DateTime.fromMillisecondsSinceEpoch(session.endTime);

      final startHour = startDate.hour + startDate.minute / 60.0;
      final endHour = endDate.hour + endDate.minute / 60.0;

      final startAngle = (startHour * 15 - 90) * pi / 180;
      final sweepAngle = (endHour - startHour) * 15 * pi / 180;

      canvas.drawArc(
        Rect.fromCircle(center: center, radius: radius - strokeWidth / 2),
        startAngle,
        sweepAngle,
        false,
        sessionPaint,
      );
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}

class _TimelineSessionItem extends StatelessWidget {
  final FocusSession session;
  const _TimelineSessionItem({required this.session});

  @override
  Widget build(BuildContext context) {
    final startStr = DateFormat('HH:mm').format(DateTime.fromMillisecondsSinceEpoch(session.startTime));
    final endStr = DateFormat('HH:mm').format(DateTime.fromMillisecondsSinceEpoch(session.endTime));

    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 8),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          SizedBox(
            width: 50,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(startStr, style: const TextStyle(fontWeight: FontWeight.bold)),
                Text(endStr, style: Theme.of(context).textTheme.bodySmall?.copyWith(color: Colors.grey)),
              ],
            ),
          ),
          const SizedBox(width: 16),
          Expanded(
            child: Card(
              color: Theme.of(context).colorScheme.surfaceContainerHighest,
              child: Padding(
                padding: const EdgeInsets.all(12),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    const Text('Focus Session', style: TextStyle(fontWeight: FontWeight.bold)),
                    Text('${session.duration ~/ 60000} minutes', style: Theme.of(context).textTheme.bodySmall),
                    if (session.note != null && session.note!.isNotEmpty)
                      Text(session.note!, style: const TextStyle(fontStyle: FontStyle.italic, fontSize: 12)),
                  ],
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class _TimelineEventItem extends StatelessWidget {
  final DailyEvent event;
  const _TimelineEventItem({required this.event});

  @override
  Widget build(BuildContext context) {
    final timeStr = DateFormat('HH:mm').format(DateTime.fromMillisecondsSinceEpoch(event.timestamp));

    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 8),
      child: Row(
        children: [
          SizedBox(width: 50, child: Text(timeStr, style: const TextStyle(fontWeight: FontWeight.bold))),
          const SizedBox(width: 16),
          Expanded(
            child: Card(
              color: Theme.of(context).colorScheme.secondaryContainer,
              child: Padding(
                padding: const EdgeInsets.all(12),
                child: Row(
                  children: [
                    const Icon(Icons.notifications, size: 16),
                    const SizedBox(width: 8),
                    Text(event.type, style: const TextStyle(fontWeight: FontWeight.bold)),
                  ],
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

File: app/src/main/java/com/github/benji377/timety/ui/screens/DailyStatsScreen.kt
package com.github.benji377.timety.ui.screens

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.github.benji377.timety.data.DailyEvent
import com.github.benji377.timety.data.FocusSession
import com.github.benji377.timety.viewmodel.StatsViewModel
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import kotlin.math.cos
import kotlin.math.sin

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun DailyStatsScreen(
    viewModel: StatsViewModel,
    initialDate: Long,
    onBack: () -> Unit
) {
    var currentDateMillis by remember { mutableLongStateOf(initialDate) }
    val allSessions by viewModel.allSessions.collectAsState()
    val user by viewModel.user.collectAsState()

    val dailySessions = remember(allSessions, currentDateMillis) {
        val calendar = Calendar.getInstance().apply { timeInMillis = currentDateMillis }
        val startOfDay = calendar.apply {
            set(Calendar.HOUR_OF_DAY, 0)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }.timeInMillis
        val endOfDay = startOfDay + 24 * 60 * 60 * 1000L
        allSessions.filter { it.startTime in startOfDay until endOfDay }
    }

    val dailyEvents by viewModel.getEventsForDay(currentDateMillis)
        .collectAsState(initial = emptyList())

    val totalFocussedMillis = dailySessions.sumOf { it.duration }
    val dailyTargetMillis = user?.dailyFocusTarget ?: (120 * 60 * 1000L)
    val progress = (totalFocussedMillis.toFloat() / dailyTargetMillis.toFloat()).coerceIn(0f, 1f)

    Scaffold(
        topBar = {
            TopAppBar(
                title = {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.Center,
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        IconButton(onClick = {
                            val cal =
                                Calendar.getInstance().apply { timeInMillis = currentDateMillis }
                            cal.add(Calendar.DAY_OF_YEAR, -1)
                            currentDateMillis = cal.timeInMillis
                        }) {
                            Icon(
                                Icons.AutoMirrored.Filled.ArrowBack,
                                contentDescription = "Previous Day"
                            )
                        }
                        Text(
                            text = SimpleDateFormat("EEEE, MMM d", Locale.getDefault()).format(
                                Date(
                                    currentDateMillis
                                )
                            ),
                            style = MaterialTheme.typography.titleMedium
                        )
                        IconButton(onClick = {
                            val cal =
                                Calendar.getInstance().apply { timeInMillis = currentDateMillis }
                            cal.add(Calendar.DAY_OF_YEAR, 1)
                            currentDateMillis = cal.timeInMillis
                        }) {
                            Icon(
                                Icons.AutoMirrored.Filled.ArrowForward,
                                contentDescription = "Next Day"
                            )
                        }
                    }
                },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                    }
                }
            )
        }
    ) { padding ->
        LazyColumn(
            modifier = Modifier
                .padding(padding)
                .fillMaxSize()
                .padding(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(24.dp)
        ) {
            item {
                Circular24hGraph(sessions = dailySessions)
            }

            item {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(
                        text = "${totalFocussedMillis / 60000} / ${dailyTargetMillis / 60000} mins",
                        style = MaterialTheme.typography.titleLarge,
                        fontWeight = FontWeight.Bold
                    )
                    Spacer(modifier = Modifier.height(8.dp))
                    LinearProgressIndicator(
                        progress = { progress },
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(8.dp),
                        strokeCap = StrokeCap.Round
                    )
                }
            }

            item {
                HorizontalDivider()
            }

            item {
                Text(
                    text = "Timeline",
                    style = MaterialTheme.typography.titleMedium,
                    modifier = Modifier.fillMaxWidth()
                )
            }

            val timelineItems = (dailySessions.map { it as Any } + dailyEvents.map { it as Any })
                .sortedBy {
                    if (it is FocusSession) it.startTime else (it as DailyEvent).timestamp
                }

            items(timelineItems) { item ->
                if (item is FocusSession) {
                    TimelineItem(item)
                } else {
                    EventTimelineItem(item as DailyEvent)
                }
            }
        }
    }
}

@Composable
fun EventTimelineItem(event: DailyEvent) {
    val timeStr = SimpleDateFormat("HH:mm", Locale.getDefault()).format(Date(event.timestamp))

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Column(modifier = Modifier.width(60.dp)) {
            Text(
                text = timeStr,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.Bold
            )
        }

        Spacer(modifier = Modifier.width(16.dp))

        Card(
            modifier = Modifier.weight(1f),
            colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.secondaryContainer)
        ) {
            Row(
                modifier = Modifier.padding(12.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Icon(
                    Icons.Default.Notifications,
                    contentDescription = null,
                    modifier = Modifier.size(16.dp)
                )
                Spacer(modifier = Modifier.width(8.dp))
                Text(
                    text = event.type,
                    style = MaterialTheme.typography.titleSmall
                )
            }
        }
    }
}

@Composable
fun Circular24hGraph(sessions: List<FocusSession>) {
    val primaryColor = MaterialTheme.colorScheme.primary
    val secondaryColor = MaterialTheme.colorScheme.outlineVariant

    Box(
        modifier = Modifier
            .size(280.dp)
            .padding(16.dp),
        contentAlignment = Alignment.Center
    ) {
        Canvas(modifier = Modifier.fillMaxSize()) {
            val center = Offset(size.width / 2, size.height / 2)
            val radius = size.minDimension / 2
            val strokeWidth = 20.dp.toPx()

            // Draw base gray circle
            drawCircle(
                color = secondaryColor,
                radius = radius,
                style = Stroke(width = strokeWidth)
            )

            // Draw hour marks
            for (i in 0 until 24) {
                val angleRad = (i * 15f - 90f) * (Math.PI / 180f).toFloat()
                val start = Offset(
                    center.x + (radius - strokeWidth / 2) * cos(angleRad),
                    center.y + (radius - strokeWidth / 2) * sin(angleRad)
                )
                val end = Offset(
                    center.x + (radius + strokeWidth / 2) * cos(angleRad),
                    center.y + (radius + strokeWidth / 2) * sin(angleRad)
                )
                drawLine(
                    color = Color.White.copy(alpha = 0.5f),
                    start = start,
                    end = end,
                    strokeWidth = 2.dp.toPx()
                )
            }

            // Draw sessions
            sessions.forEach { session ->
                val startCal = Calendar.getInstance().apply { timeInMillis = session.startTime }
                val endCal = Calendar.getInstance().apply { timeInMillis = session.endTime }

                val startHour =
                    startCal.get(Calendar.HOUR_OF_DAY) + startCal.get(Calendar.MINUTE) / 60f
                val endHour = endCal.get(Calendar.HOUR_OF_DAY) + endCal.get(Calendar.MINUTE) / 60f

                val startAngle = (startHour * 15f - 90f)
                val sweepAngle = (endHour - startHour) * 15f

                drawArc(
                    color = primaryColor,
                    startAngle = startAngle,
                    sweepAngle = sweepAngle,
                    useCenter = false,
                    style = Stroke(width = strokeWidth, cap = StrokeCap.Round)
                )
            }
        }

        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text(
                "24h",
                style = MaterialTheme.typography.headlineMedium,
                fontWeight = FontWeight.ExtraBold
            )
            Text("Focus Distribution", style = MaterialTheme.typography.bodySmall)
        }
    }
}

@Composable
fun TimelineItem(session: FocusSession) {
    val startTimeStr =
        SimpleDateFormat("HH:mm", Locale.getDefault()).format(Date(session.startTime))
    val endTimeStr = SimpleDateFormat("HH:mm", Locale.getDefault()).format(Date(session.endTime))

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Column(modifier = Modifier.width(60.dp)) {
            Text(
                text = startTimeStr,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.Bold
            )
            Text(
                text = endTimeStr,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.outline
            )
        }

        Spacer(modifier = Modifier.width(16.dp))

        Card(
            modifier = Modifier.weight(1f),
            colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceVariant)
        ) {
            Column(modifier = Modifier.padding(12.dp)) {
                Text(
                    text = "Focus Session",
                    style = MaterialTheme.typography.titleSmall
                )
                Text(
                    text = "${session.duration / 60000} minutes",
                    style = MaterialTheme.typography.bodySmall
                )
                if (!session.note.isNullOrBlank()) {
                    Text(
                        text = session.note,
                        style = MaterialTheme.typography.bodySmall,
                        fontStyle = androidx.compose.ui.text.font.FontStyle.Italic
                    )
                }
            }
        }
    }
}
